ImageBuffer subclasses should not be templates

shallawa opened the pull request:
#### c5d5c98f4d5c8ca2b2e0aa8d391467352f594256
<pre>
ImageBuffer subclasses should not be templates
<a href="https://bugs.webkit.org/show_bug.cgi?id=240058">https://bugs.webkit.org/show_bug.cgi?id=240058</a>
rdar://93085945

Reviewed by Simon Fraser.

Remove BackendType from the ImageBuffer subclasses and make ConcreteImageBuffer
hold a unique_ptr&lt;ImageBufferBackend&gt;. The create methods will be template
function. BackendType will allow creating different backends. All static and
constexpr backend members can be populated in the create method in a new type
name ImageBufferBackend::Info.

This change will allow
(1) Merging ImageBuffer and ConcreteImageBuffer
(2) Removing ImageBufferBackendSharing.
(3) Holding pointers to the subclasses RemoteImageBuffer and RemoteImageBufferProxy
(4) Moving virtaul specfic subclasses methods from ImageBuffer

* Source/WebCore/platform/graphics/ConcreteImageBuffer.h:
(WebCore::ConcreteImageBuffer::create):
(WebCore::ConcreteImageBuffer::populateBackendInfo):
(WebCore::ConcreteImageBuffer::ConcreteImageBuffer):
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::create):
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::setOwnershipIdentity):
* Source/WebCore/platform/graphics/PlatformImageBuffer.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListImageBuffer.h:
* Source/WebKit/GPUProcess/graphics/ImageBufferShareableAllocator.cpp:
(WebKit::ImageBufferShareableAllocator::createImageBuffer const):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
(WebKit::RemoteImageBuffer::create):
(WebKit::RemoteImageBuffer::RemoteImageBuffer):
(WebKit::RemoteImageBuffer::m_renderingResourcesRequest):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::createImageBufferWithQualifiedIdentifier):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::allocateBufferForBackingStore):
* Source/WebKit/Shared/WebImage.cpp:
(WebKit::WebImage::create):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/PlatformRemoteImageBufferProxy.h: Removed.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteImageBufferProxy::create):
(WebKit::RemoteImageBufferProxy::RemoteImageBufferProxy):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::createImageBuffer):
* Source/WebKit/WebProcess/GPU/graphics/ThreadSafeRemoteImageBufferFlusher.cpp: Added.
(WebKit::ThreadSafeRemoteImageBufferFlusher::ThreadSafeRemoteImageBufferFlusher):
(WebKit::ThreadSafeRemoteImageBufferFlusher::flush):
* Source/WebKit/WebProcess/GPU/graphics/ThreadSafeRemoteImageBufferFlusher.h: Renamed from Source/WebKit/GPUProcess/graphics/PlatformRemoteImageBuffer.h.
* Source/WebKit/WebProcess/GPU/webrtc/AudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::start):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::storageChanged):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::startThread):
* Source/WebKit/WebProcess/Inspector/WebInspectorUIExtensionController.cpp:
(WebKit::WebInspectorUIExtensionController::parseExtensionErrorFromEvaluationResult const):
(WebKit::WebInspectorUIExtensionController::unwrapEvaluationResultAsObject const):
(WebKit::WebInspectorUIExtensionController::createTabForExtension):
(WebKit::WebInspectorUIExtensionController::evaluateScriptForExtension):
(WebKit::WebInspectorUIExtensionController::reloadForExtension):
(WebKit::WebInspectorUIExtensionController::showExtensionTab):
(WebKit::WebInspectorUIExtensionController::evaluateScriptInExtensionTab):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createImageBuffer const):

Canonical link: <a href="https://commits.webkit.org/252266@main">https://commits.webkit.org/252266@main</a>
</pre>

smfr commented on Source/WebCore/platform/graphics/ConcreteImageBuffer.h:
```diff
@@ -77,7 +91,7 @@ class ConcreteImageBuffer : public ImageBuffer {
     void setBackend(std::unique_ptr<ImageBufferBackend>&& backend) override
```
Can this be `final` ?

smfr commented on Source/WebCore/platform/graphics/ConcreteImageBuffer.h:
```diff
@@ -330,7 +333,8 @@ class ConcreteImageBuffer : public ImageBuffer {
 
     ImageBufferBackend::Parameters m_parameters;
-    std::unique_ptr<BackendType> m_backend;
+    ImageBufferBackend::Info m_info;
```
m_backendInfo

shallawa committed:
ImageBuffer subclasses should not be templates
https://bugs.webkit.org/show_bug.cgi?id=240058
rdar://93085945

Reviewed by Simon Fraser.

Remove BackendType from the ImageBuffer subclasses and make ConcreteImageBuffer
hold a unique_ptr<ImageBufferBackend>. The create methods will be template
function. BackendType will allow creating different backends. All static and
constexpr backend members can be populated in the create method in a new type
name ImageBufferBackend::Info.

This change will allow
(1) Merging ImageBuffer and ConcreteImageBuffer
(2) Removing ImageBufferBackendSharing.
(3) Holding pointers to the subclasses RemoteImageBuffer and RemoteImageBufferProxy
(4) Moving virtaul specfic subclasses methods from ImageBuffer

* Source/WebCore/platform/graphics/ConcreteImageBuffer.h:
(WebCore::ConcreteImageBuffer::create):
(WebCore::ConcreteImageBuffer::populateBackendInfo):
(WebCore::ConcreteImageBuffer::ConcreteImageBuffer):
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::create):
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::setOwnershipIdentity):
* Source/WebCore/platform/graphics/PlatformImageBuffer.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListImageBuffer.h:
* Source/WebKit/GPUProcess/graphics/ImageBufferShareableAllocator.cpp:
(WebKit::ImageBufferShareableAllocator::createImageBuffer const):
* Source/WebKit/GPUProcess/graphics/RemoteImageBuffer.h:
(WebKit::RemoteImageBuffer::create):
(WebKit::RemoteImageBuffer::RemoteImageBuffer):
(WebKit::RemoteImageBuffer::m_renderingResourcesRequest):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::createImageBufferWithQualifiedIdentifier):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStoreCollection.mm:
(WebKit::RemoteLayerBackingStoreCollection::allocateBufferForBackingStore):
* Source/WebKit/Shared/WebImage.cpp:
(WebKit::WebImage::create):
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/PlatformRemoteImageBufferProxy.h: Removed.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
(WebKit::RemoteImageBufferProxy::create):
(WebKit::RemoteImageBufferProxy::RemoteImageBufferProxy):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::createImageBuffer):
* Source/WebKit/WebProcess/GPU/graphics/ThreadSafeRemoteImageBufferFlusher.cpp: Added.
(WebKit::ThreadSafeRemoteImageBufferFlusher::ThreadSafeRemoteImageBufferFlusher):
(WebKit::ThreadSafeRemoteImageBufferFlusher::flush):
* Source/WebKit/WebProcess/GPU/graphics/ThreadSafeRemoteImageBufferFlusher.h: Renamed from Source/WebKit/GPUProcess/graphics/PlatformRemoteImageBuffer.h.
* Source/WebKit/WebProcess/GPU/webrtc/AudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::start):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::storageChanged):
(WebKit::AudioMediaStreamTrackRendererInternalUnitManager::Proxy::startThread):
* Source/WebKit/WebProcess/Inspector/WebInspectorUIExtensionController.cpp:
(WebKit::WebInspectorUIExtensionController::parseExtensionErrorFromEvaluationResult const):
(WebKit::WebInspectorUIExtensionController::unwrapEvaluationResultAsObject const):
(WebKit::WebInspectorUIExtensionController::createTabForExtension):
(WebKit::WebInspectorUIExtensionController::evaluateScriptForExtension):
(WebKit::WebInspectorUIExtensionController::reloadForExtension):
(WebKit::WebInspectorUIExtensionController::showExtensionTab):
(WebKit::WebInspectorUIExtensionController::evaluateScriptInExtensionTab):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createImageBuffer const):

Canonical link: https://commits.webkit.org/252266@main
